src/api/viewportFeaturesBinary.js

```diff
@@ -0,0 +1,133 @@
+
+      for (let j = startIndex; j < endIndex; j++) {
+        ringCoordinates.push(
+          Array.from(
+            positions.value.subarray(j * positions.size, (j + 1) * positions.size)
+          )
+        );
+      }
+
+      if (intersects(getFeatureByType(ringCoordinates, type), viewportIntersection)) {
+        map.set(featureId, getPropertiesFromTile(data, startIndex));
+      }
+    }
+  }
+}
+
+function getIndices(data) {
+  const indices = data.primitivePolygonIndices || data.pathIndices || data.featureIds;
+  return indices.value;
+}
+
+function getFeatureId(data, startIndex) {
+  return data.featureIds.value[startIndex];
+}
+
+function getPropertiesFromTile(data, startIndex) {
+  const featureId = getFeatureId(data, startIndex);
+  const { properties, numericProps } = data;
+  const propertiesCloned = Object.assign(properties[featureId]);
+
+  for (const key in numericProps) {
+    propertiesCloned[key] = numericProps[key].value[startIndex];
+  }
+
+  return propertiesCloned;
+}
+
+function getFeatureByType(coordinates, type) {
+  switch (type) {
+    case GEOMETRY_TYPES['Polygon']:
+      return { type: 'Polygon', coordinates: [coordinates] };
+    case GEOMETRY_TYPES['LineString']:
+      return { type: 'LineString', coordinates };
+    case GEOMETRY_TYPES['Point']:
+      return { type: 'Point', coordinates: coordinates[0] };
+    default:
+      throw new Error('Invalid geometry type');
+  }
+}
+
+function calculateViewportFeatures({
+  map,
+  fullVisible,
+  viewportIntersection,
+  data,
+  type
+}) {
+  if (fullVisible) {
+    // All the features of the tile are visible
+    const indices = getIndices(data);
+
+    for (let i = 0; i < indices.length - 1; i++) {
+      const startIndex = indices[i];
+      const featureId = getFeatureId(data, startIndex);
+
+      if (!map.has(featureId)) {
+        map.set(featureId, getPropertiesFromTile(data, startIndex));
+      }
+    }
+  } else {
+    addIntersectedFeaturesInTile({ map, data, viewportIntersection, type });
+  }
+}
+
+export function viewportFeatures({ tiles, viewport }) {
+  const map = new Map();
+
+  for (const tile of tiles) {
+    // Discard if it's not a visible tile or tile has not data
+    if (!tile.isVisible || !tile.data) {
+      continue;
+    }
+
+    const { bbox } = tile;
+    const fullVisible = isTileFullVisible(bbox, viewport);
+    const viewportIntersection = bboxPolygon(prepareViewport(tile.bbox, viewport));
+
+    calculateViewportFeatures({
+      map,
+      fullVisible,
+      viewportIntersection,
+      data: tile.data.points,
+      type: GEOMETRY_TYPES['Point']
+    });
+    calculateViewportFeatures({
+      map,
+      fullVisible,
+      viewportIntersection,
+      data: tile.data.lines,
+      type: GEOMETRY_TYPES['LineString']
+    });
+    calculateViewportFeatures({
+      map,
+      fullVisible,
+      viewportIntersection,
+      data: tile.data.polygons,
+      type: GEOMETRY_TYPES['Polygon']
+    });
```

Comment:
should we avoid 3 calls with some IF or SWITCH ?

Reply:
No. We cannot know the geometry type until we access the data and I don't see the point of spending performance on that. What we can do is put these three calls in a loop to reduce lines of code

Reply:
MultiPolygon / MultiLineString... is that feasible?

Reply:
This code should be working with Multipoints, MultiLineString and MultiPolygon without changing anything